flocker/node/agents/test/test_blockdevice.py

```diff
@@ -1264,6 +1273,483 @@ def test_all_transitions(self, desired_state, discovered_state):
+                    maximum_size=int(REALISTIC_BLOCKDEVICE_SIZE.bytes),
+                    device_path=FilePath('/dev/xvdf'),
+                    mount_point=FilePath('/mount/path'),
+                )
+            ],
+            expected_datasets=[
+                DesiredDataset(
+                    state=DatasetStates.MOUNTED,
+                    dataset_id=ScenarioMixin.DATASET_ID,
+                    maximum_size=int(REALISTIC_BLOCKDEVICE_SIZE.bytes),
+                    metadata={},
+                    mount_point=self.deployer.mountroot.child(
+                        unicode(ScenarioMixin.DATASET_ID)
+                    ),
+                ),
+            ],
+            leases=Leases().acquire(
+                now=datetime.now(tz=UTC),
+                dataset_id=ScenarioMixin.DATASET_ID,
+                node_id=self.deployer.node_uuid,
+            )
+        )
+
+    def test_leased_attached_manifestaion(self):
+        """
+        If there is a lease for an attached dataset present on node,
+        there is not a corresponding desired dataset.
+        """
+        assert_desired_datasets(
+            self, self.deployer,
+            desired_manifestations=[],
+            local_datasets=[
+                DiscoveredDataset(
+                    dataset_id=ScenarioMixin.DATASET_ID,
+                    blockdevice_id=ScenarioMixin.BLOCKDEVICE_ID,
+                    state=DatasetStates.ATTACHED,
+                    maximum_size=int(REALISTIC_BLOCKDEVICE_SIZE.bytes),
+                    device_path=FilePath('/dev/xvdf'),
+                )
+            ],
+            expected_datasets=[],
+            leases=Leases().acquire(
+                now=datetime.now(tz=UTC),
+                dataset_id=ScenarioMixin.DATASET_ID,
+                node_id=self.deployer.node_uuid,
+            )
+        )
+
+    def test_leased_non_manifest(self):
+        """
+        If there is a lease for a mounted dataset present on node,
+        there is a corresponding dataset that has a desired state of
+        ``MOUNTED``.
+        """
+        assert_desired_datasets(
+            self, self.deployer,
+            desired_manifestations=[],
+            local_datasets=[
+                DiscoveredDataset(
+                    dataset_id=ScenarioMixin.DATASET_ID,
+                    blockdevice_id=ScenarioMixin.BLOCKDEVICE_ID,
+                    state=DatasetStates.NON_MANIFEST,
+                    maximum_size=int(REALISTIC_BLOCKDEVICE_SIZE.bytes),
+                )
+            ],
+            expected_datasets=[],
+            leases=Leases().acquire(
+                now=datetime.now(tz=UTC),
+                dataset_id=ScenarioMixin.DATASET_ID,
+                node_id=self.deployer.node_uuid,
+            )
+        )
+
+    def test_lease_elsewhere(self):
+        """
+        If there is a lease for a dataset on another node,
+        there isn't a corresponding dataset that has a desired state
```

Comment:
I got tripped up by this too. It has a discovered dataset, but not a desired dataset. I've adjusted the docstring slightly, but I'm open to improvements to clarify this.